packages/workbox-routing/lib/Router.mjs

```diff
@@ -0,0 +1,328 @@
+        paramName: 'event',
+      });
+    }
+
+    const url = new URL(event.request.url);
+    if (!url.protocol.startsWith('http')) {
+      _private.logger.log(`The URL '${url}' does not start with 'http', so ` +
+        `it can't be handled.`);
+      return;
+    }
+
+    let {handler, params} = this._findHandlerAndParams(event, url);
+
+    // If we don't have a handler because there was no matching route, then
+    // fall back to defaultHandler if that's defined.
+    if (!handler && this.defaultHandler) {
+      handler = this.defaultHandler;
+    }
+
+    if (handler) {
+      let responsePromise = handler.handle({url, event, params});
+      if (this.catchHandler) {
+        responsePromise = responsePromise.catch((error) => {
+          return this.catchHandler.handle({url, event, error});
+        });
+      }
+      return responsePromise;
+    }
+  }
+
+  /**
+   * Checks the incoming event.request against the registered routes, and if
+   * there's a match, returns the corresponding handler along with any params
+   * generated by the match.
+   *
+   * @param {FetchEvent} event
+   * @param {URL} url
+   * @return {Object} Returns an object with `handler` and `params` properties.
+   * They are populated if a matching route was found or `undefined` otherwise.
+   * @private
+   */
+  _findHandlerAndParams(event, url) {
+    const routes = this._routes.get(event.request.method) || [];
+    for (const route of routes) {
+      let matchResult = route.match({url, event});
+      if (matchResult) {
+        _private.logger.log(`The router found a matching route.`, {
+          route,
+          request: event.request,
+        });
+
+        if (Array.isArray(matchResult) && matchResult.length === 0) {
+          // Instead of passing an empty array in as params, use undefined.
+          matchResult = undefined;
+        } else if (matchResult.constructor === Object &&
+          Object.keys(matchResult).length === 0) {
+          // Instead of passing an empty object in as params, use undefined.
+          matchResult = undefined;
+        }
+
+        // Break out of the loop and return the appropriate values as soon as
+        // we have a match.
+        return {
+          params: matchResult,
+          handler: route.handler,
+        };
+      }
+    }
+
+    // If we didn't have a match, then return undefined values.
+    return {handler: undefined, params: undefined};
+  }
+
+  /**
+   * An optional `handler` that's called by default when no routes
+   * explicitly match the incoming request.
+   *
+   * If the default is not provided, unmatched requests will go against the
+   * network as if there were no service worker present.
+   *
+   * @param {function|module:workbox-runtime-caching.Handler} handler
+   * This parameter can be either a function or an object which is a subclass
+   * of `Handler`.
+   *
+   * Either option should result in a `Response` that the `Route` can use to
+   * handle the `fetch` event.
```

Comment:
Nit: "Either option should return a `Response` object, which will be used to respond to the `fetch` event."